File: arrays/two_pointers/3sum_closest.rb
# @param {Integer[]} nums
# @param {Integer} target
# @return {Integer}
def three_sum_closest(nums, target)
  nums.sort!
	result, n = nums[0,3].reduce(:+), nums.length - 1

  0.upto(n) do |i|
    l, r = i + 1, n
    while l < r
      s = nums[i] + nums[l] + nums[r]
      result = s if (target - s).abs <  (target - result).abs
      if s < target
        l += 1
      elsif s > target
        r -= 1
      else
        return result
      end
    end
  end

  result
end

File: two_pointers/3sum_closest.py
from typing import List


class Solution:
    def threeSumClosest(self, nums: List[int], target: int) -> int:
        nums.sort()
        n = len(nums)
        solution = sum([nums[0], nums[1], nums[2]])

        for i in range(n - 2):
            l, r = i + 1, n - 1

            while l < r:
                s = nums[i] + nums[l] + nums[r]
                if abs(target - s) < abs(target - solution): solution = s

                if s < target:      l += 1
                elif s > target:    r -= 1
                else:
                    return s

        return solution
